dra-evolution/pkg/api/claim_types.go

```diff
@@ -286,41 +343,29 @@ type AllocationResult struct {
 	// +optional
 	DriverData []DriverData `json:"driverData,omitempty" protobuf:"bytes,1,opt,name=driverData"`
 
-	// This is the node which provides the allocated devices.
+	// Setting this field is optional. If unset, the allocated devices are available everywhere.
 	//
-	// If empty, then the devices are not local to one particular node.
-	NodeName string `json:"nodeName,omitempty" protobuf:"bytes,2,opt,name=nodeName"`
+	// +optional
+	AvailableOnNodes *v1.NodeSelector `json:"availableOnNodes,omitempty" protobuf:"bytes,2,opt,name=availableOnNodes"`
```

Comment:
This is the only place where the "structured parameters" API gets changed. I am bringing back the node selector and moved the "node name" part down into `RequestAllocationResult` where it sits alongside the device name.

A simple node name is not enough because "classic DRA" is how network-attached devices can be supported in 1.31.

Reply:
In the first commit I kept the status structured as it was before. But we can go further and do what @johnbelamaric suggested and make the status more like the spec, with a list of results at the root instead of nesting those under each driver.

The second commit does that. It's not directly related to classic DRA, it just was motivated by the previous change because that one already put "nodeName" and "deviceName" into the same struct. Now that struct also has "driverName", which completes the triple that is needed to uniquely identify the device.

Reply:
I still find this confusing. I would think the default case would be "available only on the node named in the result" rather than "available everywhere". This means we have to set this for every single result, even though none in the beta will actually need it?

Actually, I am not sure this is correct even - shouldn't this be on a per request basis? Or I am not understanding something. For example, couldn't a single claim contain requests for two different network attached devices, which are available on different sets of nodes? In that case, wouldn't we need this in the request result?

I also am not clear on why we need this in the *result* at all. The result is used to tell the driver how to actuate the devices. This field here suggests that somehow a subsequent request for a device may be resolved (by the platform, not the user) to this claim? I think that's the key thing I am missing. This is used for some use case where a controller manages access to a given set of devices that are pre-allocated by a claim? If that's the case, the ReservedFor maybe makes more sense now too.

I am not sure that stuff belongs in claim.

I will say again - I think the network attached stuff is really really different, I am not sure a single API for both node-local and network attached is useful.

Reply:
> This means we have to set this for every single result, even though none in the beta will actually need it?

The beta uses this to determine which devices are allocated. For each result, it needs to connect the result back to the corresponding ResourcePool device. We do this via the `<device name>/<node name>/<device name>` triple, so that has to be stored somewhere.

That we now have `<node name>` repeated for each entry is a side effect of flattening the status. It is a bit redundant if all that the claim uses are node-local devices from the same node, but that isn't necessarily always the case. Let me rename the result field as suggested in https://github.com/kubernetes-sigs/wg-device-management/pull/26#discussion_r1630785781, then it might look better.

> I think the network attached stuff is really really different, I am not sure a single API for both node-local and network attached is useful

I see it the other way around. Node-local is a special case and with a little of "out of the box" (sic) thinking we can generalize the API so that it works for both.

Reply:
I think I am still confused as to what this node selector is for. There are two possibilities (at least), and I don't understand which is the intent:

1. Allowing us to determine the resource pool from which this result is allocated. For node-local devices, this is not needed because NodeName is in each entry. For networked devices, a node selector seems insufficiently precise; instead I would expect some identifier for the networked pool itself (if we want to allow rebalancing, it can't be the k8s name of the pool...but do we need to allow that?).
2. Given the thinking in 1), that node selector is insufficiently precise, I thought maybe you intended this to be used to "advertise" the availability of a pre-allocated claim. That's what I was trying to understand in the last comment.


It sounds, though, like maybe it is 1) after all. But I don't see how a node selector is the right thing to use as an identifier for pool of networked devices. I don't think a node selector is a good identifier.

Or am I completely misunderstanding?

Reply:
It is not 1. The purpose of the field is to tell a scheduler which sees an already allocated ResourceClaim where the devices are available. This is relevant for sharing an allocated claim between pods. Typically this is a node, but it doesn't have to be.

What identifies the device and thus the pool is the `driver name/device pool name/device name` combination, which is stored for each result. Driver and pool can be different for each result once we allow network-attached devices.

Reply:
One could argue that by looking at all devices, a scheduler can infer where they are available. But that only works when the ResourcePool currently exists, which is not required and cannot be enforced after allocation. With classic DRA, the scheduler cannot look up further information about devices and has to be given this node selector.

IMHO, an explicit field is also easier for users.

Reply:
Let me explain this differently. A claim status must answer two questions:
- Which nodes have access to the devices?
- Which devices were allocated?

In the special case of node-local devices, the answer to both was a node name. In the more general case, we need a node selector for the first question and some unique ID for a device. https://github.com/kubernetes-sigs/wg-device-management/pull/26#discussion_r1630073547 is about that.

Reply:
The was I've always thought about it is this:
1. A vendor-specific controller is the one in charge of doing the allocation of this claim
2. It may choose to make the resource available on more than one node (if possible)
3. It sets the node selector to reflect the set of nodes where the resource can/will be made available
4. The scheduler then has the freedom to schedule a pod referencing this claim on any of the nodes in the node selector (rather than just one specific node)

Reply:
Thank you, those explanations help a lot. This lgtm.